src/dmd/mars.d

```diff
@@ -2125,3 +2168,278 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+            return true; // this import will be compiled
+        }
+    }
+    return false; // this import will not be compiled
+}
+
+// A range of component identifiers for a module
+private struct ModuleComponentRange
+{
+    Identifiers* packages;
+    Identifier name;
+    bool isPackageFile;
+    size_t index;
+    @property auto totalLength() const { return packages.dim + 1 + (isPackageFile ? 1 : 0); }
+
+    @property auto empty() { return index >= totalLength(); }
+    @property auto front() const
+    {
+        if (index < packages.dim)
+            return (*packages)[index];
+        if (index == packages.dim)
+            return name;
+        else
+            return Identifier.idPool("package");
+    }
+    void popFront() { index++; }
+}
+
+/*
+ * Determines if the given module should be included in the compilation.
+ * Returns:
+ *  True if the given module should be included in the compilation.
+ */
+private bool includeImportedModuleCheck(ModuleComponentRange components)
+    in { assert(includeImports); } body
+{
+    createMatchNodes();
+    size_t nodeIndex = 0;
+    while (nodeIndex < matchNodes.dim)
+    {
+        //printf("matcher ");printMatcher(nodeIndex);printf("\n");
+        auto info = matchNodes[nodeIndex++];
+        if (info.depth <= components.totalLength())
+        {
+            size_t nodeOffset = 0;
+            for (auto range = components;;range.popFront())
+            {
+                if (range.empty || nodeOffset >= info.depth)
+                {
+                    // MATCH
+                    //printf("matcher ");printMatcher(nodeIndex - 1);
+                    //printf(" MATCHES module '");components.print();printf("'\n");
+                    return !info.isExclude;
+                }
+                if (!range.front.equals(matchNodes[nodeIndex + nodeOffset].id))
+                {
+                    break;
+                }
+                nodeOffset++;
+            }
+        }
+        //printf("matcher ");printMatcher(nodeIndex-1);
+        //printf(" does not match module '");components.print();printf("'\n");
+        nodeIndex += info.depth;
+    }
+    assert(nodeIndex == matchNodes.dim, "code bug");
+    return includeByDefault;
+}
+
+// Matching module names is done with an array of matcher nodes.
+// The nodes are sorted by "component depth" from largest to smallest
+// so that the first match is always the longest (best) match.
+private struct MatcherNode
+{
+    union
+    {
+        struct
+        {
+            ushort depth;
+            bool isExclude;
+        }
+        Identifier id;
+    }
+    this(Identifier id) { this.id = id; }
+    this(bool isExclude, ushort depth)
+    {
+        this.depth = depth;
+        this.isExclude = isExclude;
+    }
+}
+
+/*
+ * $(D includeByDefault) determines whether to include/exclude modules when they don't
+ * match any pattern. This setting changes depending on if the user provided any "inclusive" module
```

Comment:
what is meant by "pattern"?

Reply:
https://dlang.org/dmd-windows.html#switch-i[
> This behavior can be overriden by providing patterns via -i=<pattern>. A pattern of the form -i=<package> is an "inclusive pattern", whereas a pattern of the form -i=-<package> is an "exclusive pattern". Inclusive patterns will include all module's whose names match the pattern, whereas exclusive patterns will exclude them. For example. all modules in the package foo.bar can be included using -i=foo.bar or excluded using -i=-foo.bar. Note that each component of the fully qualified name must match the pattern completely, so the pattern foo.bar would not match a module named foo.barx.